### DISCUSSION TITLE
[Feature proposal] Alternative autoplay engine support

### DISCUSSION THREAD

**Discussion by SonoSooS** [ppy/osu#18031] [Ideas]
(In reference to #14635)

## Preface
The default autoplay engine works, and has stood the test of time, but it has its limitations.

While the default EaseOut direct path interpolation between two hit circles, and direct slider follow does technically produce a valid autoplay, on some technical maps (that is, where you have to play smarter, not harder) it's not at all representative of a real player's movements, and in some cases the replay produced by the autoplay is impossible to perform by a human[1].
This can be a problem for beginner players (or for even some intermediate players trying to learn harder beatmaps), who are using replays as a way of observing cursor movements, and trying to deduct practices from it, or to learn anything at all.

On the other side of the spectrum, there are players who want to watch autoplays for fun/entertainment/relaxation, but the default autoplay is just very stiff and boring. There is also no configuration to change any of the behavior, like the interpolation algorithm for example.

[1] For example:
- spam sliders get perfectly followed in the dead center of the follow ring, while real players just rest the cursor in the middle of the slider until the spam slider is over
- some movements are nearly impossible, or completely impossible to be performed by a human as done by the autoplay, where the autoplay just changes direction instantly, while humans need to redirect the momentum into a different direction by steering the movement vector

## Plan
The high level overview on how to add this functionality into the game, is that in the mod config slideout panel ("Customize" button in the mod selector popout) to have a dropdown box, from where we can select an autoplay engine.

After selecting the autoplay engine, the settings for the selected autoplay engine would appear below the dropdown box.

## Proposal
Unless it's already possible/implemented, first a new mechanism needs to be added to the mod config slideout panel, which would allow extra configuration to be exposed at runtime depending on what options are selected. There are two ways I could think of to implement this:
- Allow configurable objects to report sub-configurable objects, effectively creating a categorically sorted hierarchy tree
  - Problem: it either has to be recursive, or it has to be flattened if we want to use this mechanism elsewhere
- Allow exposing configs dynamically upon config update
  - Problem: really error-prone, has to use hardcoded strings, and requires explicit type checking

Cinema and Autoplay mod sources also have to be either slightly refactored so they can share the same autoplay selection/configuration/generation code, or there has to be a helper class created to which we can offload the logistics of the newly added functionality.

## Problems
This change can't be added to Autopilot, as not only would it change the difficulty curve of competitive Autopilot play, but it would also open a vulnerability to the server-side anticheat code, as some autoplay engines are so configurable, that it's technically possible for each user to submit an Autopilot play with completely unique mouse movement, meaning it would be ultra easy to fake it with external programs.

Due to how Cinema and Autoplay mods are handled, it also reqiures some unavoidable refactoring, to prevent code copypaste as much as possible, while still letting the Framework do its magic without having to resort to dirty hacks and workarounds.

Depending on how we want to implement the dynamic configurability of the autoplay engines, it might need changes to osu!Framework, meaning it's possible to create weird behavior if the commits pushed to osu!Framework don't match the commits pushed to the game's repository.

**Comment by bdach** [comment]
First off, I have no idea how you've bypassed the issue template, but don't do this please.

Secondly, this discussion doesn't really need to exist. It could as well have been continued in the old cursor dance thread (https://github.com/ppy/osu/discussions/14635#discussioncomment-2665751).

**Reply by SonoSooS** [reply]
I am terribly sorry. I just did what's stated in CONTRIBUTING.md, and took it literally.
> If you wish to work on something with the intention of having it included in the official distribution, >please open an issue< for discussion and we will give you what you need from a design perspective to proceed.

I used the "Reference in new issue" option.
![image](https://user-images.githubusercontent.com/7823099/166112613-b50293d8-5941-442d-99a3-b7908829a1fa.png)


(I am also unable to post into the arrival channel on the dev server, because I am unable to verify my phone number, so this is the option I chose, since I had doubts of a new role being created just for me so I phone verification could be bypassed.)

**Reply by bdach** [reply]
Yes well that particular quoted snippet predates discussions being enabled in this repository. I may see to it being updated..

**Comment by peppy** [comment]
I am sure your intentions here are good, but most of the issue text is confusing or baffling.

Are you wanting to attempt this task? If so I'd just go for it.

**Reply by bdach** [reply]
I believe the giant wall of text essentially boils down to whether we'd permit a setting for autoplay that changes the type of movement used to allow implementing cursor dancing and that sort of thing, which I'd be fine with. But there's also talk of settings enabling/disabling other settings and I wouldn't go down that route. I'd start with a few predetermined presets and not make every single tiny detail configurable (yet?)

**Reply by SonoSooS** [reply]
I was actually *trying* to avoid presets, as that's what I already have, and I find it ultra limiting. Although going with the preset route means much less work needed to port over existing code (except for the vigorous refactor needed for my existing old code).

I guess most players don't actually care about the name of which autoplay engine is being used, so I could get away with a single dropdown box with each preset having an one- or two-word description, like "Default", "Playful", "Player-like", "Rule breaker", etc.

**Reply by SonoSooS** [reply]
Some features of some presets need the preemption time and circle radius, but during my research on how to do it, I ended up confusing myself even more.

Eventually I figured out that I can just use OsuHitObject.TimePreempt and OsuHitObject.Radius.

My question is, is the valid way to use these values is to check them for each hit object, or do I have to calculate them based on time using a different way?
If this is a valid way to use them, is the Radius a constant during that one hit object (mostly considering sliders here), or do I have to take radius fluctiations into account as well?

Also, do I have to worry about other difficulty-changing mods explicitly (like DifficultyAdjust or Daycore), or do I only have to worry about IApplicableToRate mods (like in OsuAutoGeneratorBase)? Which values do I need to apply these values to, if to any, besides spinner spin rate?

**Reply by frenzibyte** [reply]
> My question is, is the valid way to use these values is to check them for each hit object, or do I have to calculate them based on time using a different way?
If this is a valid way to use them, is the Radius a constant during that one hit object (mostly considering sliders here), or do I have to take radius fluctiations into account as well?

Hit objects by default have a constant radius and scale adjusted by `BeatmapDifficutly.CircleSize`, but eventually, [users may have variable circle size across the beatmap](https://github.com/ppy/osu/discussions/13028). It is best to stay on one constant size for simplicity.

> Also, do I have to worry about other difficulty-changing mods explicitly (like DifficultyAdjust or Daycore), or do I only have to worry about IApplicableToRate mods (like in OsuAutoGeneratorBase)? Which values do I need to apply these values to, if to any, besides spinner spin rate?

"Difficulty adjust" should have its effect on `BeatmapDifficulty` before autoplay is generated. "Daycore" should be part of `IApplicableToRate` mods, in which the clock speed matters for the replay (I suppose?).

**Comment by ramjam18** [comment]
The nerd that I am, I've actually wanted to be able to configure Autoplay a little more than you can at the moment for a long time now, myself. This is partially because I really like osu!'s automation features, and I also like playing with Autopilot if I want to sit back and relax a bit.

The problem is, I know only enough coding to make a stopwatch console app, so I'm afraid I cannot exactly weigh into this discussion.

That said, while I think cursordancing Autoplay would be a bit unnecessary, even though it'd be entertaining to see this pulled off, I'd actually also like to voice my support for a little bit of configurability.


Slightly offtopic PS: I've found that changing Easing.Out to Easing.InOutCirc works _wonders_ for playability with Autopilot. Not to mention, it actually also looks fairly realistic as well.


EDIT: Again slightly offtopic, but I just felt like [messing around a little](https://www.youtube.com/watch?v=xkpZKaqCRhY). (Video also showcases Easing.InOutCirc.)
